src/index.js

```diff
@@ -104,168 +119,191 @@ class ExtractTextPlugin {
                         meta.content,
                         module,
                         extractedChunk
                       );
                     }
-                    callback();
                   }
-                } else callback();
-              },
-              (err) => {
-                if (err) return callback(err);
-                callback();
-              }
-            );
-          },
-          (err) => {
-            if (err) return callback(err);
-            extractedChunks.forEach((extractedChunk) => {
-              if (isInitialOrHasNoParents(extractedChunk)) {
-                this.mergeNonInitialChunks(extractedChunk);
-              }
-            }, this);
-            extractedChunks.forEach((extractedChunk) => {
-              if (!isInitialOrHasNoParents(extractedChunk)) {
-                extractedChunk.forEachModule((module) => {
-                  extractedChunk.removeModule(module);
-                });
+
+                  moduleCallback();
+                },
+                (err) => {
+                  if (err) {
+                    return chunkCallback(err);
+                  }
+
+                  chunkCallback();
+                }
+              );
+            },
+            (err) => {
+              if (err) {
+                return callback(err);
               }
-            });
-            compilation.applyPlugins(
-              'optimize-extracted-chunks',
-              extractedChunks
-            );
-            callback();
-          }
-        );
-      });
-      compilation.plugin('additional-assets', (callback) => {
+
+              extractedChunks.forEach((extractedChunk) => {
+                if (isInitialOrHasNoParents(extractedChunk)) {
+                  this.mergeNonInitialChunks(extractedChunk);
+                }
+              }, this);
+
+              extractedChunks.forEach((extractedChunk) => {
+                if (!isInitialOrHasNoParents(extractedChunk)) {
+                  for (const chunkModule of extractedChunk.modulesIterable) {
+                    extractedChunk.removeModule(chunkModule);
+                  }
+                }
+              });
+
+              compilation.hooks.optimizeExtractedChunks.call(extractedChunks);
+              callback();
+            }
+          );
+        }
+      );
+
+      compilation.hooks.additionalAssets.tapAsync(plugin.name, (assetCb) => {
         extractedChunks.forEach((extractedChunk) => {
           if (extractedChunk.getNumberOfModules()) {
-            extractedChunk.sortModules((a, b) => {
-              if (!options.ignoreOrder && isInvalidOrder(a, b)) {
-                compilation.errors.push(
-                  new OrderUndefinedError(a.getOriginalModule())
-                );
-                compilation.errors.push(
-                  new OrderUndefinedError(b.getOriginalModule())
-                );
-              }
-              return getOrder(a, b);
```

Comment:
The getOrder helper function uses the underlying index2 property of modules. Unfortunately this property seems to be set by webpack after [sealing](https://webpack.js.org/api/compilation/#seal) and apparently sealing happens after the [additionalAssets](https://webpack.js.org/api/compilation/#additional-assets-async) hook, which is where the sorting takes place, because accessing index2 yielded null every time.

Reply:
Please add a comment for this to the code :)